{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f96e957a-680f-4e45-b172-b325c55b249c",
   "metadata": {},
   "source": [
    "# Preprocessing sequence data\n",
    "\n",
    "The first step in the pipeline is the sequence preprocessing, which usually consinsts of two main steps:\n",
    "  * Segmentation\n",
    "  * Tokenization\n",
    "\n",
    "The segmentation is illustrated in figure <img src=\"https://github.com/nbrg-ppcu/prokbert/blob/main/assets/Figure2_segmentation.png?raw=true\" width=\"500\" height=\"300\" alt=\"Segmentation Process\"> .\n",
    "\n",
    "The genomic language models (glm) only able to process a limited sized chunk of sequence data at once, which is typcailly 0-4kb long, thus first it is needed to split it into multiple parts. This process is called segmentation. \n",
    "The segmentation can be contiguous, splitting the sequence into disjunct segments and 'random', randomly sampling segments with length of L. \n",
    "The first step is loading the sequence from fasta file. Many cases reasenable to add the reverse complement of sequence s needed as well. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d14bef41-1fb3-4991-8004-a263726a5c7a",
   "metadata": {},
   "source": [
    "# Preprocessing Sequence Data\n",
    "\n",
    "The initial stage of our pipeline involves two primary steps: segmentation and tokenization.\n",
    "\n",
    "Segmentation is depicted in the figure below. Genomic language models (GLMs) process limited-size chunks of sequence data, typically ranging from 0 to 4kb. Therefore, it's essential to divide the sequence into smaller parts, a process known as segmentation. Segmentation can be either contiguous, which splits the sequence into disjoint segments, or random, which involves randomly sampling segments of length L. \n",
    "\n",
    "The first practical step in segmentation is loading the sequence from a FASTA file. Often, it's also beneficial to include the reverse complement of the sequence.\n",
    "Segmentation process:\n",
    "\n",
    "<img src=\"https://github.com/nbrg-ppcu/prokbert/blob/main/assets/Figure2_segmentation.png?raw=true\" width=\"500\" height=\"300\" alt=\"Segmentation Process\"> \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "148f54d8-1ee9-43e7-8d3c-8947dc2518f7",
   "metadata": {},
   "source": [
    "# Loading the sequence data \n",
    "\n",
    "The `load_contigs` function efficiently handles sequence data from FASTA files, providing options to include reverse complements and sequence metadata. When the output is set to a DataFrame (`AsDataFrame=True`), the function organizes this data into a structured format, enhancing data accessibility and manipulation for downstream analyses.\n",
    "\n",
    "The resulting DataFrame consists of the following columns:\n",
    "- `sequence_id`: Unique identifier of the sequence (integer) if `is_add_sequence_id=True`.\n",
    "- `fasta_id`: identifier of the sequence, parsed from the fasta file.\n",
    "- `description`: Description or metadata associated with the sequence, typically extracted from the FASTA file.\n",
    "- `source_file`: Path of the source FASTA file.\n",
    "- `sequence`: Nucleotide sequence. Sequences are converted to uppercase if `to_uppercase=True`.\n",
    "- `orientation`: Indicates 'forward' for original sequences and 'reverse' for reverse complements if `adding_reverse_complement=True`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4fddeeae-0083-4f49-b2d2-9db53a18b510",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-11-11 21:42:04,684 - INFO - Loading sequence data into memory!\n",
      "2023-11-11 21:42:04,685 - INFO - Since the fasta_files_list is a string, not a list, we convert it to a list.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_id</th>\n",
       "      <th>fasta_id</th>\n",
       "      <th>description</th>\n",
       "      <th>source_file</th>\n",
       "      <th>sequence</th>\n",
       "      <th>orientation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>NC_000913.3_2524612_2527387</td>\n",
       "      <td>NC_000913.3_2524612_2527387 NC_000913.3 Escher...</td>\n",
       "      <td>data/ESKAPE_sample.fasta</td>\n",
       "      <td>TATATTAGAAATGTCCGCGACCTTTCATACATACCACCGGTACGCC...</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NC_000913.3_2524612_2527387</td>\n",
       "      <td>NC_000913.3_2524612_2527387 NC_000913.3 Escher...</td>\n",
       "      <td>data/ESKAPE_sample.fasta</td>\n",
       "      <td>CCAGCAATGGTGAAAGGAAAATCCCCAGCAGGCTGGATGCCGACGC...</td>\n",
       "      <td>reverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>NC_007795.1_1702041_1705080</td>\n",
       "      <td>NC_007795.1_1702041_1705080 NC_007795.1 Staphy...</td>\n",
       "      <td>data/ESKAPE_sample.fasta</td>\n",
       "      <td>ATAATGCTAAATCGTAACCCCACTGCTTAAATGAGCCTTCTGTAAA...</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>NC_007795.1_1702041_1705080</td>\n",
       "      <td>NC_007795.1_1702041_1705080 NC_007795.1 Staphy...</td>\n",
       "      <td>data/ESKAPE_sample.fasta</td>\n",
       "      <td>GAATGGACTGTTATCAGGATCAACTTGCTGCCAAGGGATAATAGAC...</td>\n",
       "      <td>reverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>NC_002516.2_410266_413817</td>\n",
       "      <td>NC_002516.2_410266_413817 NC_002516.2 Pseudomo...</td>\n",
       "      <td>data/ESKAPE_sample.fasta</td>\n",
       "      <td>GTGACCGGGGTCAGGTTCTCGGCGGCGGCGCGCATCACGTGCTTGC...</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>NC_002516.2_410266_413817</td>\n",
       "      <td>NC_002516.2_410266_413817 NC_002516.2 Pseudomo...</td>\n",
       "      <td>data/ESKAPE_sample.fasta</td>\n",
       "      <td>GCGTTTCTACCGCGAGGTCGGCCCACTCGACTGCACCCTGGAGAGT...</td>\n",
       "      <td>reverse</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sequence_id                     fasta_id  \\\n",
       "0            0  NC_000913.3_2524612_2527387   \n",
       "1            1  NC_000913.3_2524612_2527387   \n",
       "2            2  NC_007795.1_1702041_1705080   \n",
       "3            3  NC_007795.1_1702041_1705080   \n",
       "4            4    NC_002516.2_410266_413817   \n",
       "5            5    NC_002516.2_410266_413817   \n",
       "\n",
       "                                         description  \\\n",
       "0  NC_000913.3_2524612_2527387 NC_000913.3 Escher...   \n",
       "1  NC_000913.3_2524612_2527387 NC_000913.3 Escher...   \n",
       "2  NC_007795.1_1702041_1705080 NC_007795.1 Staphy...   \n",
       "3  NC_007795.1_1702041_1705080 NC_007795.1 Staphy...   \n",
       "4  NC_002516.2_410266_413817 NC_002516.2 Pseudomo...   \n",
       "5  NC_002516.2_410266_413817 NC_002516.2 Pseudomo...   \n",
       "\n",
       "                source_file  \\\n",
       "0  data/ESKAPE_sample.fasta   \n",
       "1  data/ESKAPE_sample.fasta   \n",
       "2  data/ESKAPE_sample.fasta   \n",
       "3  data/ESKAPE_sample.fasta   \n",
       "4  data/ESKAPE_sample.fasta   \n",
       "5  data/ESKAPE_sample.fasta   \n",
       "\n",
       "                                            sequence orientation  \n",
       "0  TATATTAGAAATGTCCGCGACCTTTCATACATACCACCGGTACGCC...     forward  \n",
       "1  CCAGCAATGGTGAAAGGAAAATCCCCAGCAGGCTGGATGCCGACGC...     reverse  \n",
       "2  ATAATGCTAAATCGTAACCCCACTGCTTAAATGAGCCTTCTGTAAA...     forward  \n",
       "3  GAATGGACTGTTATCAGGATCAACTTGCTGCCAAGGGATAATAGAC...     reverse  \n",
       "4  GTGACCGGGGTCAGGTTCTCGGCGGCGGCGCGCATCACGTGCTTGC...     forward  \n",
       "5  GCGTTTCTACCGCGAGGTCGGCCCACTCGACTGCACCCTGGAGAGT...     reverse  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from prokbert.sequtils import *\n",
    "\n",
    "fasta_file = 'data/ESKAPE_sample.fasta'\n",
    "# This pandas dataframe holds the sequence data parsed from the fasta file\n",
    "sequences = load_contigs(fasta_file,IsAddHeader=True, adding_reverse_complement=True, AsDataFrame=True, to_uppercase=True, is_add_sequence_id=True)\n",
    "\n",
    "sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b464ef58-1057-47ba-b6fd-9363f2e683ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9b126631-6f46-4073-9c22-4288965d90c9",
   "metadata": {},
   "source": [
    "# Segmentation and tokenization\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "444b2c1e-916a-451c-809e-66492724d3a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-11-11 21:43:41,138 - INFO - Checking input DataFrame!\n",
      "2023-11-11 21:43:41,139 - INFO - Checking input sequence_id is valid primary key in the DataFrame\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'segment': 'ATCGATCG', 'segment_start': 0, 'segment_end': 8, 'sequence_id': nan}, {'segment': 'AAATTTTT', 'segment_start': 8, 'segment_end': 16, 'sequence_id': nan}]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>segment_id</th>\n",
       "      <th>sequence_id</th>\n",
       "      <th>segment_start</th>\n",
       "      <th>segment_end</th>\n",
       "      <th>segment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>512</td>\n",
       "      <td>TATATTAGAAATGTCCGCGACCTTTCATACATACCACCGGTACGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>512</td>\n",
       "      <td>1024</td>\n",
       "      <td>TTATGCTATGAAAAAACATCTTTTAACTCTGACACTTTCCTCTATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1024</td>\n",
       "      <td>1536</td>\n",
       "      <td>CCGGCTGGGTTGCAGGTTACAACTTTATGCTGGGCAGCGAGAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1536</td>\n",
       "      <td>2048</td>\n",
       "      <td>CCGATGCCTGCGGCTACCATCGGGAACAGCGTCGCCGGATGTCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2048</td>\n",
       "      <td>2560</td>\n",
       "      <td>AGCCAGCTGCTGCCCTGCATCTGTAAGCACCACTTCACGCGTGGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2560</td>\n",
       "      <td>2780</td>\n",
       "      <td>ACCTTCGTGCTGTTTCCGATTCTGGGTGTACTGTTTGCCTGGTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>512</td>\n",
       "      <td>CCAGCAATGGTGAAAGGAAAATCCCCAGCAGGCTGGATGCCGACGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>512</td>\n",
       "      <td>1024</td>\n",
       "      <td>CCATTCATGAAGAAAACAGATGAATTATTCTTTAAAACAATTAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>1024</td>\n",
       "      <td>1536</td>\n",
       "      <td>AGCCTTTCTTTCTGCTTTGCCATCGCGATAGCGCTTTGGCCGTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>1536</td>\n",
       "      <td>2048</td>\n",
       "      <td>TAAAAATCTTCGCCCAGTTTGTCATCCGCATATCGATACTGAATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>2048</td>\n",
       "      <td>2560</td>\n",
       "      <td>CATTTCTCCCCAACTAAAGTTAGCGCCACCTTCAAATTCAAGATAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>2560</td>\n",
       "      <td>2780</td>\n",
       "      <td>AAGACTTATAAACCTGATTTCACGCCACGAGTGGCCTTTATTTTAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>512</td>\n",
       "      <td>ATAATGCTAAATCGTAACCCCACTGCTTAAATGAGCCTTCTGTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>512</td>\n",
       "      <td>1024</td>\n",
       "      <td>GTTGTATCAAATGCTTTTTGGCCAGCTAGCACTTCTTTCCATTCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>1024</td>\n",
       "      <td>1536</td>\n",
       "      <td>TTCAAATAATTCTTCACGACCAGCGTCTTTCGTAATTTGACGGCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>1536</td>\n",
       "      <td>2048</td>\n",
       "      <td>CATAACGATTTTCATCTGATTCATTTTCAGCATCAGGATCGAAATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>2048</td>\n",
       "      <td>2560</td>\n",
       "      <td>TTAAAGCTTTTTCAAGATGTGAATAATCCATGAATTTTACAATAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>2560</td>\n",
       "      <td>3039</td>\n",
       "      <td>TGGCAAAATATGTTAGCTTTTGGTTCCCAGAAATTCCAAACTGGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>512</td>\n",
       "      <td>GAATGGACTGTTATCAGGATCAACTTGCTGCCAAGGGATAATAGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>3</td>\n",
       "      <td>512</td>\n",
       "      <td>1024</td>\n",
       "      <td>AACAGAACCAGTATGTCCATCCGGTAACAAATCCTGCAAAAGGCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>1024</td>\n",
       "      <td>1536</td>\n",
       "      <td>ATAGATTTACAAAAAAATTGTTGCACATATAATGCCAGTTTTTATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "      <td>1536</td>\n",
       "      <td>2048</td>\n",
       "      <td>CATTAGTTACAGCGTTTGCTCGAGTAAGACAAGATAAAGAACCACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>3</td>\n",
       "      <td>2048</td>\n",
       "      <td>2560</td>\n",
       "      <td>GAAAAAGGATTAATTCCTAATGTTGATTTTTATAGTGCGAGTGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>3</td>\n",
       "      <td>2560</td>\n",
       "      <td>3039</td>\n",
       "      <td>AATTAAAGAATATTTAATTGCTGTTAAAGGACCTTTAACAACACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>512</td>\n",
       "      <td>GTGACCGGGGTCAGGTTCTCGGCGGCGGCGCGCATCACGTGCTTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>4</td>\n",
       "      <td>512</td>\n",
       "      <td>1024</td>\n",
       "      <td>GAACAGCAGGTCGCGCAGGGCCTTCAGCCACTGCACGCGCTGCGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>4</td>\n",
       "      <td>1024</td>\n",
       "      <td>1536</td>\n",
       "      <td>GCGAGCGCAGCGTCACCACCAACCATATCGCCACCCACCTGGGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>4</td>\n",
       "      <td>1536</td>\n",
       "      <td>2048</td>\n",
       "      <td>GGGCACATCGAGGCGCAGGCCCAAGATGCGGTGGACGGGATTTCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>4</td>\n",
       "      <td>2048</td>\n",
       "      <td>2560</td>\n",
       "      <td>GGACGCCGGCCAGCTCAGTGCTGCCGCAGGCGGGCCTCCAGCCAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>4</td>\n",
       "      <td>2560</td>\n",
       "      <td>3072</td>\n",
       "      <td>CACCGACACCGCGACGCCGCCGAGCAGCGGGCAGTCCCCAGCGGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>30</td>\n",
       "      <td>4</td>\n",
       "      <td>3072</td>\n",
       "      <td>3551</td>\n",
       "      <td>GGGCGGTGGTGGTGTGGCCACCAGCTTACCAGCCGAACTCGTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>31</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>512</td>\n",
       "      <td>GCGTTTCTACCGCGAGGTCGGCCCACTCGACTGCACCCTGGAGAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>32</td>\n",
       "      <td>5</td>\n",
       "      <td>512</td>\n",
       "      <td>1024</td>\n",
       "      <td>AGCCCGCCTGGTGGCTGCCCAATCCGCACCTGCAGACCCTGTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>33</td>\n",
       "      <td>5</td>\n",
       "      <td>1024</td>\n",
       "      <td>1536</td>\n",
       "      <td>GGCCTGGGCTTTTCCCGGGTCTACCAGGCGCACTTCATGAAGGCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>34</td>\n",
       "      <td>5</td>\n",
       "      <td>1536</td>\n",
       "      <td>2048</td>\n",
       "      <td>GAAAAAGGGGGGCGAATGGAACTGACGCTGCGGCACAACCCGATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35</td>\n",
       "      <td>5</td>\n",
       "      <td>2048</td>\n",
       "      <td>2560</td>\n",
       "      <td>ACGCAGTTGCGCCTCGCTGAGGTCGCCGTCGAGATCGAGGGTGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>36</td>\n",
       "      <td>5</td>\n",
       "      <td>2560</td>\n",
       "      <td>3072</td>\n",
       "      <td>CGATGCGATCGCGGGTCTTGATGCGTGGTGCCATGGGGACGACGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>37</td>\n",
       "      <td>5</td>\n",
       "      <td>3072</td>\n",
       "      <td>3551</td>\n",
       "      <td>CGGGACTTCAGCGCGCGCTCCGCCGACGAGACCCTGCTCGCCGAGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    segment_id  sequence_id  segment_start  segment_end  \\\n",
       "0            0            0              0          512   \n",
       "1            1            0            512         1024   \n",
       "2            2            0           1024         1536   \n",
       "3            3            0           1536         2048   \n",
       "4            4            0           2048         2560   \n",
       "5            5            0           2560         2780   \n",
       "6            6            1              0          512   \n",
       "7            7            1            512         1024   \n",
       "8            8            1           1024         1536   \n",
       "9            9            1           1536         2048   \n",
       "10          10            1           2048         2560   \n",
       "11          11            1           2560         2780   \n",
       "12          12            2              0          512   \n",
       "13          13            2            512         1024   \n",
       "14          14            2           1024         1536   \n",
       "15          15            2           1536         2048   \n",
       "16          16            2           2048         2560   \n",
       "17          17            2           2560         3039   \n",
       "18          18            3              0          512   \n",
       "19          19            3            512         1024   \n",
       "20          20            3           1024         1536   \n",
       "21          21            3           1536         2048   \n",
       "22          22            3           2048         2560   \n",
       "23          23            3           2560         3039   \n",
       "24          24            4              0          512   \n",
       "25          25            4            512         1024   \n",
       "26          26            4           1024         1536   \n",
       "27          27            4           1536         2048   \n",
       "28          28            4           2048         2560   \n",
       "29          29            4           2560         3072   \n",
       "30          30            4           3072         3551   \n",
       "31          31            5              0          512   \n",
       "32          32            5            512         1024   \n",
       "33          33            5           1024         1536   \n",
       "34          34            5           1536         2048   \n",
       "35          35            5           2048         2560   \n",
       "36          36            5           2560         3072   \n",
       "37          37            5           3072         3551   \n",
       "\n",
       "                                              segment  \n",
       "0   TATATTAGAAATGTCCGCGACCTTTCATACATACCACCGGTACGCC...  \n",
       "1   TTATGCTATGAAAAAACATCTTTTAACTCTGACACTTTCCTCTATA...  \n",
       "2   CCGGCTGGGTTGCAGGTTACAACTTTATGCTGGGCAGCGAGAAATT...  \n",
       "3   CCGATGCCTGCGGCTACCATCGGGAACAGCGTCGCCGGATGTCCAA...  \n",
       "4   AGCCAGCTGCTGCCCTGCATCTGTAAGCACCACTTCACGCGTGGTT...  \n",
       "5   ACCTTCGTGCTGTTTCCGATTCTGGGTGTACTGTTTGCCTGGTGGA...  \n",
       "6   CCAGCAATGGTGAAAGGAAAATCCCCAGCAGGCTGGATGCCGACGC...  \n",
       "7   CCATTCATGAAGAAAACAGATGAATTATTCTTTAAAACAATTAAAG...  \n",
       "8   AGCCTTTCTTTCTGCTTTGCCATCGCGATAGCGCTTTGGCCGTGGA...  \n",
       "9   TAAAAATCTTCGCCCAGTTTGTCATCCGCATATCGATACTGAATCC...  \n",
       "10  CATTTCTCCCCAACTAAAGTTAGCGCCACCTTCAAATTCAAGATAA...  \n",
       "11  AAGACTTATAAACCTGATTTCACGCCACGAGTGGCCTTTATTTTAG...  \n",
       "12  ATAATGCTAAATCGTAACCCCACTGCTTAAATGAGCCTTCTGTAAA...  \n",
       "13  GTTGTATCAAATGCTTTTTGGCCAGCTAGCACTTCTTTCCATTCAA...  \n",
       "14  TTCAAATAATTCTTCACGACCAGCGTCTTTCGTAATTTGACGGCTC...  \n",
       "15  CATAACGATTTTCATCTGATTCATTTTCAGCATCAGGATCGAAATG...  \n",
       "16  TTAAAGCTTTTTCAAGATGTGAATAATCCATGAATTTTACAATAGA...  \n",
       "17  TGGCAAAATATGTTAGCTTTTGGTTCCCAGAAATTCCAAACTGGAT...  \n",
       "18  GAATGGACTGTTATCAGGATCAACTTGCTGCCAAGGGATAATAGAC...  \n",
       "19  AACAGAACCAGTATGTCCATCCGGTAACAAATCCTGCAAAAGGCCC...  \n",
       "20  ATAGATTTACAAAAAAATTGTTGCACATATAATGCCAGTTTTTATA...  \n",
       "21  CATTAGTTACAGCGTTTGCTCGAGTAAGACAAGATAAAGAACCACT...  \n",
       "22  GAAAAAGGATTAATTCCTAATGTTGATTTTTATAGTGCGAGTGTTT...  \n",
       "23  AATTAAAGAATATTTAATTGCTGTTAAAGGACCTTTAACAACACCA...  \n",
       "24  GTGACCGGGGTCAGGTTCTCGGCGGCGGCGCGCATCACGTGCTTGC...  \n",
       "25  GAACAGCAGGTCGCGCAGGGCCTTCAGCCACTGCACGCGCTGCGTG...  \n",
       "26  GCGAGCGCAGCGTCACCACCAACCATATCGCCACCCACCTGGGTAT...  \n",
       "27  GGGCACATCGAGGCGCAGGCCCAAGATGCGGTGGACGGGATTTCCG...  \n",
       "28  GGACGCCGGCCAGCTCAGTGCTGCCGCAGGCGGGCCTCCAGCCAGT...  \n",
       "29  CACCGACACCGCGACGCCGCCGAGCAGCGGGCAGTCCCCAGCGGTT...  \n",
       "30  GGGCGGTGGTGGTGTGGCCACCAGCTTACCAGCCGAACTCGTCGGC...  \n",
       "31  GCGTTTCTACCGCGAGGTCGGCCCACTCGACTGCACCCTGGAGAGT...  \n",
       "32  AGCCCGCCTGGTGGCTGCCCAATCCGCACCTGCAGACCCTGTGGAG...  \n",
       "33  GGCCTGGGCTTTTCCCGGGTCTACCAGGCGCACTTCATGAAGGCCA...  \n",
       "34  GAAAAAGGGGGGCGAATGGAACTGACGCTGCGGCACAACCCGATGC...  \n",
       "35  ACGCAGTTGCGCCTCGCTGAGGTCGCCGTCGAGATCGAGGGTGGTG...  \n",
       "36  CGATGCGATCGCGGGTCTTGATGCGTGGTGCCATGGGGACGACGGG...  \n",
       "37  CGGGACTTCAGCGCGCGCTCCGCCGACGAGACCCTGCTCGCCGAGA...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from prokbert.config_utils import *\n",
    "# This class provide validated input parameters for segmentation and tokenization\n",
    "defconfig = SeqConfig() # For the detailed configarion parameters see: https://github.com/nbrg-ppcu/prokbert/blob/main/src/prokbert/configs/sequence_processing.yaml\n",
    "\n",
    "segmentation_params = {'max_length' : 8, # We split the sequence into L  \n",
    "                        'min_length' : 3,\n",
    "                         'type' : 'contiguous'} #default segmentation type\n",
    "tokenization_params = {}\n",
    "\n",
    "segmentation_params = defconfig.get_and_set_segmentation_parameters(segmentation_params)\n",
    "tokenization_params = defconfig.get_and_set_tokenization_parameters(tokenization_params)\n",
    "\n",
    "# Segment single sequence:\n",
    "segment_list = segment_sequence_contiguous('ATCGATCGAAATTTTTT', segmentation_params)\n",
    "print(segment_list)\n",
    "\n",
    "# Segment a DataFrame:\n",
    "segmentation_params = {'max_length' : 512, # We split the sequence into L  \n",
    "                        'min_length' : 6,\n",
    "                         'type' : 'contiguous'} #default segmentation type\n",
    "segmentdb = segment_sequences(sequences, segmentation_params, AsDataFrame=True)\n",
    "segmentdb\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#segmentation_params = defconfig.get_and_set_segmentation_parameters()  \n",
    "#tokenization_params = defconfig.get_and_set_tokenization_params()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c09ef812-11bc-4de9-b1be-585cae7d7d16",
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {'min_length': 0, 'max_length': 100}\n",
    "segment_sequence_contiguous('ATCGATCGA', params)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8122ac53-d77d-4ab4-9bf1-d7b37d9844cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "segmentation_params\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6fc423e-78bf-43c4-bebd-27004681f199",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
